edb/ir/typeutils.py

```diff
@@ -305,39 +301,46 @@ def _typeref(
         )
     elif not isinstance(t, s_types.Collection):
         assert isinstance(t, s_types.InheritingType)
-        union_of = t.get_union_of(schema)
-        union: Optional[FrozenSet[irast.TypeRef]]
-        if union_of:
-            assert isinstance(t, s_objtypes.ObjectType)
-            union_types = {
-                cast(s_objtypes.ObjectType, c).get_nearest_non_derived_parent(
-                    schema)
-                for c in union_of.objects(schema)
-            }
-            non_overlapping, union_is_exhaustive = (
-                s_utils.get_non_overlapping_union(schema, union_types)
+
+        union: Optional[FrozenSet[irast.TypeRef]] = None
+        union_is_exhaustive: bool = False
+        ancestors: Optional[FrozenSet[irast.TypeRef]] = None
+        children: Optional[FrozenSet[irast.TypeRef]] = None
+        if t.is_union_type(schema) or t.is_intersection_type(schema):
+            normalized: s_utils.NormalizedTypeExpr = (
+                s_utils.NormalizedTypeExpr.create(t, schema=schema)
+            )
+
+            non_overlapping_descendants, union_is_exhaustive = (
+                normalized.get_non_overlapping_descendants(schema=schema)
+            )
+
+            non_overlapping = cast(
+                FrozenSet[s_objtypes.ObjectType],
+                non_overlapping_descendants
             )
             if union_is_exhaustive:
                 non_overlapping = frozenset({
-                    t for t in non_overlapping
-                    if t.is_material_object_type(schema)
+                    c for c in non_overlapping
+                    if c.is_material_object_type(schema)
                 })
 
             union = frozenset(
                 _typeref(c) for c in non_overlapping
             )
-        else:
-            union_is_exhaustive = False
-            union = None
-
-        intersection_of = t.get_intersection_of(schema)
-        intersection: Optional[FrozenSet[irast.TypeRef]]
-        if intersection_of:
-            intersection = frozenset(
-                _typeref(c) for c in intersection_of.objects(schema)
+
+            # Always include children and ancestors for complex types
+            ancestor_types = normalized.get_ancestors(schema=schema)
+            child_types = normalized.get_descendants(schema=schema)
```

Comment:
Is this new behavior?

Reply:
Yes. The issue is in `_find_rvar_in_intersection_by_typeref` and `type_contains`. When compiling the intersection, the typeref is used to find an appropriate rvar using the concept of "containment". With only simple unions or intersections this is straightforward enough.

When dealing with an intersection with potentially type expressions on either side, figuring out whether a typeref "contains" another becomes harder since the schema is not available.

For example, with `R` = `(A | B) & C` has the ancestors `C` and `Obj`. However, `L` = `A | B` should count as "containing" `R` despite not being a direct ancestor. The approach taken here to solve this is to check that `L.children` is a superset of `R.children`.

Reply:
More cases detailed in `type_contains`